### DISCUSSION TITLE
Is it ok to call getNow() on an EspSntpClock object

### DISCUSSION THREAD

**Discussion by chrish987** [bxparks/AceTimeClock#24] [Q&A]
More a question than an issue.

I am using ESP32 modules and using the EspSntpClock object to create an NTP clock. The NTP server is local. I am then using ntpclock.getNow() in lots of places through my code to get current timestamps. Some of these calls are in callbacks that are supposed to complete quickly.

I read somewhere in the docs recently that this is a blocking call. I have not seen any issues with my code, nor noticed any impacts of blocking.

Is what I am doing ok, or should I be creating a systemClock object from the EspSntpClock and then calling the systemClock.getNow() to get timestamps?

Does the EspSntpClock.getNow() call out to the NTP server on every call?

**Comment by bxparks** [comment]
The `Clock::getNow()` is intended to be a blocking call. However, the `EspSntpClock` subclass implements `getNow()` by essentially just calling the `time(nullptr)` function from the C library of the ESP32. I am unable to find the source code for `time()` on the ESP32, but I believe it calls the underlying Espressif IDF layer. I have no idea where the source code for that would be located. It is likely that the ESP32 C library's version of `time()` keeps an internal buffer, which is incremented using an interrupt, and the `time()` function call just returns the current value of the buffer. Somewhere else, if I remember, the ESP32 SNTPClient makes periodic requests to the NTP server to keep the clock in sync. Basically, the ESP32 library has implemented a version of my SystemClock, keeping the ESP32's clock synchronized in a background thread. So the `time()` function on the ESP32 should return almost instantly, and the fact that `ClocK::getNow()` is a blocking call doesn't matter.

The class where the blocking nature of `Clock::getNow()` matters is in the `NtpClock` class, whose `getNow()` method actually makes a network request to an NTP server. I have seen that network request take multiple seconds sometimes, because the DNS resolution can take a long time. For `NtpClock`, it's important to use the non-blocking API through the `SystemClock`. Because of that, on the ESP32 and ESP8266, it's probably easiest to just use the `EspSntpClock`.

It's been a long time since I have looked at this code, so I don't know if I remember all the relevant details. Take a look at a previous discussion #6 where I seemed to have discussed some of this with someone else. Maybe my previous self was better informed than my current self.

Hope this helps.

**Reply by chrish987** [reply]
thankyou. given i have seen no issue it is likely you are correct.
